trillian/ctdns/ctdns_server/main.go

```diff
@@ -0,0 +1,250 @@
+		// Regularly update the internal STH for each log so our metrics stay up-to-date with any tree head
+		// changes.
+		for _, c := range cfg.LogConfigs.Config {
+			ticker := time.NewTicker(*getSTHInterval)
+			go func(c *configpb.LogConfig) {
+				glog.Infof("start internal get-sth operations on log %v (%d)", c.Prefix, c.LogId)
+				for t := range ticker.C {
+					glog.V(1).Infof("tick at %v: force internal get-sth for log %v (%d)", t, c.Prefix, c.LogId)
+					if _, err := ctfe.GetTreeHead(ctx, clientMap[c.LogBackendName], c.LogId, c.Prefix); err != nil {
+						glog.Warningf("failed to retrieve tree head for log %v (%d): %v", c.Prefix, c.LogId, err)
+					}
+				}
+			}(c)
+		}
+	}
+
+	// Bring up the DNS udpServer and serve until we get a signal not to.
+	go awaitSignal(func() {
+		os.Exit(1)
+	})
+	// TODO(Martin2112): Might need a separate metrics endpoint like CTFE.
+	// Bring up the UDP udpServer and allow time for it to start.
+	ch := make(chan bool)
+	udpServer := dns.Server{
+		Addr:       *metricsEndpoint,
+		Net:        "udp",
+		TsigSecret: nil,
+		NotifyStartedFunc: func() {
+			ch <- true
+		},
+	}
+
+	// Allow a short time for the UDP server to notify us that it is ready.
+	go udpServer.ListenAndServe()
+	select {
+	case res := <-ch:
+		glog.Infof("UDP udpServer has started OK: %v", res)
+	case <-time.After(*startupWait):
+		glog.Fatal("UDP udpServer not listening within timeout")
+	}
+
+	// Now start the TCP server.
+	tcpServer := dns.Server{
+		Addr:       *metricsEndpoint,
+		Net:        "tcp",
+		TsigSecret: nil}
+	if err := tcpServer.ListenAndServe(); err != nil {
+		glog.Errorf("Failed to setup the TCP DNS Server: %s\n", err.Error())
+	}
+	glog.Flush()
+}
+
+// awaitSignal waits for standard termination signals, then runs the given
+// function; it should be run as a separate goroutine.
+func awaitSignal(doneFn func()) {
+	// Arrange notification for the standard set of signals used to terminate a server
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+
+	// Now block main and wait for a signal
+	sig := <-sigs
+	glog.Warningf("Signal received: %v", sig)
+	glog.Flush()
+
+	doneFn()
+}
+
+func setupDNSHandler(client trillian.TrillianLogClient, deadline time.Duration, cfg *configpb.LogConfig) error {
+	opts := ctfe.InstanceOptions{Deadline: deadline, MetricFactory: prometheus.MetricFactory{}, RequestLog: new(ctfe.DefaultRequestLog)}
+	handler := ctdns.New(client, cfg, opts)
+	dns.DefaultServeMux.Handle(cfg.DnsZone, handler)
+	return nil
+}
+
+func readMultiCfg(filename string) (*configpb.LogMultiConfig, error) {
+	cfg, err := ctfe.MultiLogConfigFromFile(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	return cfg, nil
+}
+
+func readCfg(filename string, backendSpec string) (*configpb.LogMultiConfig, error) {
+	cfg, err := ctfe.LogConfigFromFile(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	return ctfe.ToMultiLogConfig(cfg, backendSpec), nil
+}
```

Comment:
Lots of copy-pasted code between here and ct_server/main.go -- drop a TODO to commonize?

Reply:
Done.